### PR TITLE
Giving Sample on library v21

### PR DESCRIPTION
**Pull request by jlengrand**

### CODE REVIEW

File: giving-example/README.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+3. Visit [http://localhost:8080/](http://localhost:8080/) pick the Card component integration, follow the instructions, and after the payment you will be offered to make a donation.
+
+To try out this application with test card numbers, visit [Test card numbers](https://docs.adyen.com/development-resources/test-cards/test-card-numbers).
+We recommend saving some test cards in your browser so you can test your integration faster in the future.
+
+## Testing webhooks
+
+Webhooks deliver asynchronous notifications and it is important to test them during the setup of your integration. You can find more information about webhooks in [this detailed blog post](https://www.adyen.com/blog/Integrating-webhooks-notifications-with-Adyen-Checkout).
+
+This sample application provides a simple webhook integration exposed at `/api/webhooks/notifications`. For it to work, you need to:
```
Comment (gcatanese) [inline giving-example/README.md:87]: Mention the other Giving webhook as well

File: giving-example/src/main/java/com/adyen/giving/api/CheckoutResource.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        if(applicationProperty.getApiKey() == null) {
+            log.warn("ADYEN_KEY is UNDEFINED");
+            throw new RuntimeException("ADYEN_KEY is UNDEFINED");
+        }
+
+        var client = new Client(applicationProperty.getApiKey(), Environment.TEST);
+        this.paymentsApi = new PaymentsApi(client);
+    }
+
+    @PostMapping("/donations")
```
Comment (gcatanese) [inline giving-example/src/main/java/com/adyen/giving/api/CheckoutResource.java:48]: Add javadoc

File: giving-example/src/main/java/com/adyen/giving/api/CheckoutResource.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+     * @throws ApiException from Adyen API.
+     */
+    @PostMapping("/initiatePayment")
+    public ResponseEntity<PaymentResponse> payments(@RequestHeader String host, @RequestBody PaymentRequest body, HttpServletRequest request) throws IOException, ApiException {
+        var paymentRequest = new PaymentRequest();
+
+        var orderRef = UUID.randomUUID().toString();
+        var amount = new Amount()
+            .currency("EUR")
+            .value(10000L); // value is 10€ in minor units
```
Comment (gcatanese) [inline giving-example/src/main/java/com/adyen/giving/api/CheckoutResource.java:103]: Correct comment, it is 100EUR

File: giving-example/src/main/java/com/adyen/giving/api/GivingWebhookResource.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+        // fetch first (and only) NotificationRequestItem
+        var notificationRequestItem = notificationRequest.getNotificationItems().stream().findFirst();
+
+            if (notificationRequestItem.isPresent()) {
+
+            var item = notificationRequestItem.get();
+
+            try {
+                if (new HMACValidator().validateHMAC(item, this.applicationProperty.getHmacKey())) {
```
Comment (gcatanese) [inline giving-example/src/main/java/com/adyen/giving/api/GivingWebhookResource.java:53]: Declare and inject HMACValidator bean

File: giving-example/src/main/java/com/adyen/giving/web/CheckoutController.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @GetMapping("/")
+    public String index() {
+        return "index";
+    }
+
+    @GetMapping("/preview")
+    public String preview(@RequestParam String type, Model model) {
+        model.addAttribute("type", type);
```
Comment (gcatanese) [inline giving-example/src/main/java/com/adyen/giving/web/CheckoutController.java:36]: We pass `type` to the views but it is not necessary (we have one payment option so we can hardcode the div id in the view - as it is done for the donation DIV). I suggest to remove type everywhere (controller, view, jS)

File: giving-example/src/main/resources/static/adyenGiving.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      paymentMethodsConfiguration: {
+        ideal: {
+          showImage: true,
+        },
+        card: {
+          hasHolderName: true,
+          holderNameRequired: true,
+          name: "Credit or debit card",
+          amount: {
+            value: 10000,  // in minor units
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenGiving.js:54]: Complete comment

File: giving-example/src/main/resources/static/adyenGiving.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    // backgroundUrl: "https://www.adyen.com/dam/jcr:38701562-9572-4aae-acd3-ed53e220d63b/social-responsibility-city-illustration.svg",
+    description: "The Charitable Foundation is a non-profit aiming at showing you the power of Adyen Giving",
+    logoUrl: "https://www.adyen.com/dam/jcr:49277359-f3b5-4ceb-b54c-08189ae2433e/hands-rockon-icon-green.svg",
+    name: "The Charitable Foundation",
+    url: "https://www.adyen.com/social-responsibility/giving",
+    showCancelButton: true,
+    disclaimerMessage: {
+      message: "By donating you agree to the %#terms%#",
+      linkText: "terms and conditions",
+      link: "https://www.adyen.com/terms-and-conditions"
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenGiving.js:76]: this link gets 404: maybe it is best here to use a page within the sample app with a dummy/empty Terms and Conditions page

File: giving-example/src/main/resources/static/adyenImplementation.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
+const clientKey = document.getElementById("clientKey").innerHTML;
+const type = document.getElementById("type").innerHTML;
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenImplementation.js:2]: remove type as suggested above

File: giving-example/src/main/resources/static/adyenImplementation.js [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+      paymentMethodsConfiguration: {
+        ideal: {
+          showImage: true,
+        },
+        card: {
+          hasHolderName: true,
+          holderNameRequired: true,
+          name: "Credit or debit card",
+          amount: {
+            value: 1000,
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenImplementation.js:22]: Correct amount

File: giving-example/src/main/resources/static/adyenImplementation.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            currency: "EUR",
+          },
+        },
+        paypal: {
+          amount: {
+            value: 1000,
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenImplementation.js:28]: Correct amount

File: giving-example/src/main/resources/static/adyenImplementation.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+      onSubmit: (state, component) => {
+        if (state.isValid) {
+          handleSubmission(state, component, "/api/initiatePayment");
+        }
+      },
+      onAdditionalDetails: (state, component) => {
+        handleSubmission(state, component, "/api/submitAdditionalDetails");
+      },
+    };
+    // `spring.jackson.default-property-inclusion=non_null` needs to set in
```
Comment (gcatanese) [inline giving-example/src/main/resources/static/adyenImplementation.js:47]: I find this comment a bit confusing, I would remove it and add a comment in `application.properties` to describe what the property is about

File: giving-example/src/main/resources/templates/checkout.html [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+      xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
+      layout:decorate="~{layout}">
+<head>
+  <title>Adyen Giving demo</title>
+</head>
+<body>
+<div layout:fragment="content" id="payment-page">
+  <!-- Hidden divs with data passed from the Express server -->
+  <div id="clientKey" class="hidden" th:text="${clientKey}"></div>
+  <div id="type" class="hidden" th:text="${type}"></div>
```
Comment (gcatanese) [inline giving-example/src/main/resources/templates/checkout.html:12]: Remove type as suggested above

File: giving-example/src/main/resources/templates/index.html [hunk 13]
```diff
@@ -0,0 +1,41 @@
+<html lang="en"
+      xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
+      layout:decorate="~{layout}">
+<head>
+  <title>Select type</title>
```
Comment (gcatanese) [inline giving-example/src/main/resources/templates/index.html:5]: Update title to Adyen Giving demo

File: giving-example/src/main/resources/templates/layout.html [hunk 14]
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE html>
+<html
+  lang="en"
+  xmlns:th="http://www.thymeleaf.org"
+  xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
+>
+<head>
+  <title layout:title-pattern="$LAYOUT_TITLE - $CONTENT_TITLE">
+    Checkout Demo
```
Comment (gcatanese) [inline giving-example/src/main/resources/templates/layout.html:9]: Update title to Adyen Giving demo. Note sometimes we hardcode it, sometimes we use placeholders `$LAYOUT_TITLE`. We should make it consistent

File: giving-example/src/main/resources/templates/result.html [hunk 15]
```diff
@@ -0,0 +1,34 @@
+<html lang="en"
+      xmlns:th="http://www.thymeleaf.org"
+      xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
+      layout:decorate="~{layout}">
+<head>
+  <title th:text="${type}"></title>
```
Comment (gcatanese) [inline giving-example/src/main/resources/templates/result.html:6]: We should rename this to `result`, but I see we use it already in the other samples. Maybe we can create a TechDebt backlog, I find it confusing.

**Review by gcatanese — CHANGES_REQUESTED** [review]
Other remarks:

After making a payment the `success.svg` is just below (attached) MyStore logo, add some space to make it tidier

.gitpod.yml: add giving example
dependabot.yml add giving example

Great job, it is a good and clear example.

File: giving-example/src/main/java/com/adyen/giving/web/CheckoutController.java (hunk 5 above)
Comment (jlengrand) [reply giving-example/src/main/java/com/adyen/giving/web/CheckoutController.java:36]: You have sharp eyes btw, I appreciate that

File: giving-example/src/main/resources/static/adyenImplementation.js (hunk 11 above)
Comment (jlengrand) [reply giving-example/src/main/resources/static/adyenImplementation.js:47]: I don't even know if this is still relevant to be honest, it comes from the other sample

File: giving-example/src/main/resources/templates/result.html (hunk 15 above)
Comment (jlengrand) [reply giving-example/src/main/resources/templates/result.html:6]: What should we rename to result, the parameter "type"?

**Comment by jlengrand** [comment]
* Adyen Giving sample skeleton
* Adds docker image
* Add README
* Adding logic for webhook
* Add 'giving' workflow for E2E and update page title
* Process feedback

**Pull request merged**